---
title: "1_loading_filtering_normalizing"
author: "Pedro"
date: "8/9/2022"
output: html_document
editor_options: 
  chunk_output_type: console
---

# 0 - Load libraries
the renv package will handle package isntallation for us. it can take a while, specially for the strinli package. If you are using R 4.1.2 if should run quite smoothly!
```{r}

# install all pakages recoded in the lockfile
renv::restore()

#load necessary libraries
library(phyloseq)
library(ggplot2)
library(vegan)

```



# 1 - Load biom and fasta files in R
```{r}
# This will load all the essential data (OTU frequencies, representative sequences, mapping files with metadata, taxonomy, phylogenetic tree) for a single phyloseq object
physeq<-import_biom(BIOMfilename = "./Data/16S_phyloseq_input_FeatureTable_metadata_taxonomy.biom",
                    refseqfilename ="./Data/16S_dna-sequences.fasta",
                    version = 2.1)

# check the data
head(otu_table(physeq))
head(sample_data(physeq))
head(tax_table(physeq))
head(refseq(physeq))

```


# 2 - change ASV names

```{r}

# note that dada2 will give us horrible OTU names, let's change this with a custom function
source("./Code/Functions/backup_and_rename.R") # source will load the function inside the Code fodler of this R project

#this will run the custom fucntion and update the physeq object
physeq<-backup_and_rename(physeq) # this function runs slow inside the chuck but fast on the console.... to solve this, click on the gear and set "chuck output in the console" instead of "chunk output inline"


```


# 3 - check and remove plant sequences
```{r}

# First, a quick check to detect the presence of  o__Chloroplast or f__Mitochondria
plot_bar(subset_taxa(physeq, Order=="o__Chloroplast")) # this shows we do have some plastid DNA arround
plot_bar(subset_taxa(physeq, Family=="f__Mitochondria")) # this shows we do have some mitochndrial DNA arround


#load and run  the fucntion that will remove the plant DNA. press F2 after selecting the custom function to open it on a new tab
source("./Code/Functions/remove_Chloroplast_Mitochondria.R") 
physeq_clean<- remove_Chloroplast_Mitochondria(physeq)

# This will check if you still have those taxa in your phyloseq object. if the output is FALSE, then you got rid of them
"o__Chloroplast" %in% tax_table(physeq_clean)
"f__Mitochondria" %in% tax_table(physeq_clean)

```


# 4 - filter rare taxa & bad samples
```{r}
# check library sizes
sort(sample_sums(physeq_clean))
hist(sort(sample_sums(physeq_clean)))


# to be valid, an ASV may not be a singleton
physeq_filtered<-physeq_clean#... but first let's save the non-filtered phyloseq object in separate
otu_table(physeq_filtered) <- otu_table(physeq_filtered)[which (rowSums(otu_table(physeq_filtered)) > 7),] #this drops ~5000 taxa. 

#let's see how many sequences were kept after this filtering
sum(sample_sums(physeq_filtered))/sum(sample_sums(physeq_clean))*100 

# we kept 99.2% of sequences. let's now compare the histograms
ggplot() + 
  geom_density(aes(x=sample_sums(physeq_filtered)),fill=3, alpha=0.5) + 
  geom_density(aes(x=sample_sums(physeq_clean)),fill=1, alpha=0.5) 


#OF6 and OF1 are completetly unusable, so elts remove them
physeq_filtered<-subset_samples(physeq = physeq_filtered, sample_sums(physeq_filtered)>1000)

# there is a huge difference in sampling depth to the next 4 wosrse sequenicng librarie. with this, we have a very good sequencing but lose 6 sampels
physeq_goodSamples<-subset_samples(physeq = physeq_filtered, sample_sums(physeq_filtered)>40000)




```

# 5 - check rarefaction curve
6 samples are way worse than the other 40 samples. keeping those would compromise the amount of diversity we can really see in the dataset
```{r}


#draw rarefaction curve
# here we want to find a plateau: despite increase in the number of DNA reads, we do not increase the number of observed species. essentially, sequencing was deep enought to saturate your sampling effort.
rarecurve(t(as.data.frame(otu_table(physeq_filtered))), 
          label = FALSE, 
          step = 200,
          main="Rarefaction at 1022 reads, 7+ occurences", ylab = "Number of ASVs", xlab = "Number of DNA reads",
          abline(v = min(sample_sums(physeq_filtered)), col="red", lwd=3, lty=2))

# now let's check the good samples
rarecurve(t(as.data.frame(otu_table(physeq_goodSamples))), 
          label = FALSE, 
          step = 200,
          main="Rarefaction at 4000 reads, 7+ occurences", ylab = "Number of ASVs", xlab = "Number of DNA reads",
          abline(v = min(sample_sums(physeq_goodSamples)), col="red", lwd=3, lty=2))


#let's save this ploy that shows all your samples & the rarefaction choice
rarecurve(t(as.data.frame(otu_table(physeq_filtered))), 
          label = FALSE, 
          step = 200,
          main="Rarefaction at 4000 reads, 7+ occurences, 6 of 46 samples removed", ylab = "Number of ASVs", xlab = "Number of DNA reads",
          abline(v = min(sample_sums(physeq_goodSamples)), col="red", lwd=3, lty=2))



```

# 6 normalize dataset
For aplha diversity, we should use rarefied data, with the deepenst rarefaction we can have while losing the elast amount fo samples as possible
for beta diversity we can use CSS normalization instead

## 6a - rarefaction
```{r}

# now, let's rarefy the data
set.seed(100) # set a random seed so that whenever you re-run this code you draw the same set of OTUs
physeq_goodSamples_rarefied <- rarefy_even_depth(physeq_goodSamples, 
                                             sample.size = min(sample_sums(physeq_goodSamples)), # rarefy at minimum library size
                                             rngseed = FALSE,
                                             replace = TRUE, 
                                             trimOTUs = TRUE,
                                             verbose = TRUE) 


```



## 6b - CSS normalization
```{r}


# first, let's transform the phyloseq object into an MR experiment object
MRexp_objt<-phyloseq_to_metagenomeSeq(physeq_goodSamples)

# normalizes the object by cummulative sum scaling, a widely used method
cumNorm(MRexp_objt) 

#here you can acess the abundance matrix normalized by cummulative sum scaling. you could overwirte the phyloseq object with this
CSS_matrix <- MRcounts(MRexp_objt, norm = TRUE, log = TRUE) # using a log scale will essentially reduce the impact of common species and increase the impact of rare species

#make a new phyloseq object...
physeq_goodSamples_CSS<-physeq_goodSamples

# and now change it's taxa table
otu_table(physeq_goodSamples_CSS)<-otu_table(CSS_matrix, taxa_are_rows = TRUE)

#this is your final phyloseq object
physeq_goodSamples_CSS

```



# 7 - export data for other scripts
Now that the data has been filtered and processed, we can save it as a separate R files to be loaded in other scripts

```{r}
# first remove unecessary objects
rm(MRexp_objt, physeq, physeq_clean)


# these is out output
       physeq_filtered                # unnormalized data, with 44 samples, including library sizes of 1k to 5k reads
       physeq_goodSamples             # unnormalized data, with 40 samples, including library sizes of 40K reads or more
       physeq_goodSamples_rarefied    # rarefied data, useful for alpha diversity and neutral models
       physeq_goodSamples_CSS         # CSS-normalized data, useful for beta diversity
     
# now save them externally
save.image(file = "./Code/phyloseq_objects.RData") # R file that contains all relevant data        

```

